src/client/cli/commands/archive.js

```diff
@@ -0,0 +1,145 @@
+// @flow
+
+const fs = require('fs')
+const zlib = require('zlib')
+const tar = require('tar-stream')
+const {subcommand} = require('../util')
+const {Statement} = require('../../../model/statement')
+import type {RestClient} from '../../api'
+
+const STMT_BATCH_SIZE = 1024
+const OBJECT_BATCH_SIZE = 1024
+const TAR_ENTRY_OPTS = {
+  uid: 500,
+  gid: 500,
+  uname: 'mediachain',
+  gname: 'staff'
+}
+
+function leftpad (str, length, char = '0') {
+  return char.repeat(Math.max(0, length - str.length)) + str
+}
+
+module.exports = {
+  command: 'archive <queryString>',
+  description: 'Create a gzipped tar archive of the statements and data objects returned for the given `queryString`',
+  builder: {
+    output: {
+      alias: 'o',
+      type: 'string',
+      description: 'Filename to output archive tarball to.  If not given archive will be written to stdout.',
+      required: false,
+      default: null
+    }
+  },
+  handler: subcommand((opts: {client: RestClient, queryString: string, output?: ?string}) => {
+    const {client, queryString} = opts
+    let outputStream
+    const tarball = tar.pack()
+    const gzip = zlib.createGzip()
+    const objectIds: Set<string> = new Set()
+
+    return client.queryStream(queryString)
+      .then(response => new Promise((resolve, reject) => {
+        const {output} = opts
+        const queryStream = response.stream()
+        const outStreamName = output || 'standard output'
+        outputStream = (output == null) ? process.stdout : fs.createWriteStream(output)
+        tarball.pipe(gzip).pipe(outputStream)
+
+        let stmtBatch: Array<string> = []
+        let stmtBatchNumber = 0
+        function writeStatementBatch (force: boolean = false) {
+          if (force || stmtBatch.length >= STMT_BATCH_SIZE) {
+            const content = Buffer.from(stmtBatch.join('\n'), 'utf-8')
+            const filename = `stmt/${leftpad(stmtBatchNumber.toString(), 8)}.ndjson`
```

Comment:
just `.json` should work too for the extension? it's common for .json files to actually contain ndjson.